anymail/backends/mailtrap.py

```diff
@@ -0,0 +1,248 @@
+    ):
+        assert recipient_type in ["to", "cc", "bcc"]
+        self.data.setdefault(recipient_type, []).append(self._mailtrap_email(email))
+
+    def set_subject(self, subject):
+        self.data["subject"] = subject
+
+    def set_reply_to(self, emails):
+        self.unsupported_feature("Mailtrap does not support reply_to")
+
+    def set_extra_headers(self, headers):
+        self.data.setdefault("headers", {}).update(headers)
+
+    def set_text_body(self, body):
+        self.data["text"] = body
+
+    def set_html_body(self, body):
+        if "html" in self.data:
+            # second html body could show up through multiple alternatives,
+            # or html body + alternative
+            self.unsupported_feature("multiple html parts")
+        self.data["html"] = body
+
+    def add_attachment(self, attachment: Attachment):
+        att: MailtrapAttachment = {
+            "filename": attachment.name or "",
+            "content": attachment.b64content,
+        }
+        if attachment.mimetype:
+            att["type"] = attachment.mimetype
+        if attachment.inline:
+            att["disposition"] = "inline"
+            att["content_id"] = attachment.cid
+        self.data.setdefault("attachments", []).append(att)
+
+    def set_metadata(self, metadata):
+        self.data.setdefault("custom_variables", {}).update(
+            {str(k): str(v) for k, v in metadata.items()}
+        )
+        self.metadata = metadata  # save for set_merge_metadata
+
+    def set_template_id(self, template_id):
+        # Mailtrap requires integer (not string) TemplateID:
+        self.data["template_id"] = template_id
+
+    def set_merge_data(self, merge_data):
+        self.data.setdefault("template_variables", {}).update(merge_data)
+
+    def set_merge_global_data(self, merge_global_data):
+        self.data.setdefault("template_variables", {}).update(merge_global_data)
+
+    def set_esp_extra(self, extra):
+        update_deep(self.data, extra)
+
+
+class EmailBackend(AnymailRequestsBackend):
+    """
+    Mailtrap API Email Backend
+    """
+
+    esp_name = "Mailtrap"
+
+    def __init__(self, **kwargs):
+        """Init options from Django settings"""
+        self.api_token = get_anymail_setting(
+            "api_token", esp_name=self.esp_name, kwargs=kwargs, allow_bare=True
+        )
+        api_url = get_anymail_setting(
+            "api_url",
+            esp_name=self.esp_name,
+            kwargs=kwargs,
+            default="https://send.api.mailtrap.io/api/",
+        )
+        if not api_url.endswith("/"):
+            api_url += "/"
+
+        test_api_url = get_anymail_setting(
+            "test_api_url",
+            esp_name=self.esp_name,
+            kwargs=kwargs,
+            default="https://sandbox.api.mailtrap.io/api/",
+        )
+        if not test_api_url.endswith("/"):
+            test_api_url += "/"
+        self.test_api_url = test_api_url
+
+        bulk_api_url = get_anymail_setting(
```

Comment:
I haven't quite determined how Backends should handle bulk emails, but it also does not appear that Mailtrap's APIs are any different for their "Email sending" vs "Bulk email" APIs https://api-docs.mailtrap.io/docs/mailtrap-api-docs/.

Maybe this `self.bulk_api_url` is unnecessary, and it should be up to the user to configure which API url to use via `self.api_url`?

Reply:
Yeah, it looks like Mailtrap's ["bulk" vs transactional stream](https://help.mailtrap.io/article/113-sending-streams) is sort of a choice of outbound server pool. (Postmark and Amazon SES have similar concepts. They allow choosing the stream at send time by, respectively, providing a different API key or a different "configuration set" payload option.)

> Maybe this `self.bulk_api_url` is unnecessary, and it should be up to the user to configure which API url to use via `self.api_url`?

I think that's correct. Make transactional the default api_url, and users who want some sends to go through the bulk API can override it. (Globally via settings, or for specific sends using [`get_connection()`][1]. Plus there's an upcoming Django "EMAIL_PROVIDERS" feature that will make this even easier.)

[1]: https://anymail.dev/en/stable/installation/#anymail-settings-reference:~:text=you%20can%20override%20most%20settings%20on%20a%20per%2Dinstance%20basis%20by%20providing%20keyword%20args%20where%20the%20instance%20is%20initialized%20(e.g.%2C%20in%20a%20get_connection()%20call%20to%20create%20an%20email%20backend%20instance

Reply:
Removed `bulk_api_url`